scripts/policy/frameworks/analyzer/packet-segment-logging.zeek

```diff
@@ -0,0 +1,50 @@
+##! This script enables logging of packet segment data when a protocol
+##! parsing violation is encountered.  The amount of data from the
+##! packet logged is set by the :zeek:see:`Analyzer::Logging::packet_segment_size` variable.
+##! A caveat to logging packet data is that in some cases, the packet may
+##! not be the packet that actually caused the protocol violation.
+
+module Analyzer::Logging;
+
+export {
+	redef record connection += {
+		## A chunk of the payload that most likely resulted in a
+		## analyzer violation.
+		packet_segment: string &optional &log;
+	};
+
+	redef record Analyzer::Logging::Info += {
+		## A chunk of the payload that most likely resulted in the
+		## analyzer violation.
+		packet_segment: string &optional &log;
+	};
+
+	## Size of the packet segment to display in the DPD log.
+	option packet_segment_size: int = 255;
+}
+
+# stash the packet segment in the event causing the violation, so that it can be retrieved later.
+event analyzer_violation_info(atype: AllAnalyzers::Tag, info: AnalyzerViolationInfo) &priority=4
+	{
+	if ( ! is_protocol_analyzer(atype) && ! is_packet_analyzer(atype) )
+		return;
+
+	if ( ! info?$c || ! info?$aid )
+		return;
+
+	info$c$packet_segment = fmt("%s", sub_bytes(get_current_packet()$data, 0, packet_segment_size));
+	}
+
+hook Analyzer::Logging::log_policy(rec: Analyzer::Logging::Info, id: Log::ID, filter: Log::Filter)
```

Comment:
Interesting.

What do you think about an explicit policy hook in `Analyzer::Logging` similar to what is in `Notice` to enable this? Or feel like that is weird/old-style? To me the reliance on the generic log policy stream seems a bit weird.

E.g.
```
module Analyzer::Logging;
    # Customization point for logging.
    global policy: hook(rec: Analyzer::Logging::Info, info: AnalyzerViolationInfo);
```

This way could also use `info$c` directly, rather than the lookups.

Reply:
I'm a bit torn on that. Not needing a lookup for the connection seems like a plus. On the other hand - this kind of scenario seems to be exactly what the log_policy hooks should be here for.

Having to add a secondary hook for the cases where we actually want to modify the logs in a policy seems weird to me.